### PR TITLE
feat: add api to get first aborted uploads in series of continuous aborts

### PR DESCRIPTION
**Pull request by itsmihir**
# Description

API to get the first aborted uploads in a series of continuous aborts. For example, let's say the last 5 uploads to a destination are aborted (let's say the uploads have ids: 4, 5, 6, 7, 8) then the API will return the first aborted upload i.e the upload with id 4.

## Linear Ticket

https://linear.app/rudderstack/issue/OBS-500/get-warehouses-last-sync-status-for-listing-active-warehouse-sync

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/api/grpc.go [hunk 1]
```diff
@@ -969,3 +969,32 @@ func statsInterceptor(statsFactory stats.Stats) grpc.UnaryServerInterceptor {
...
 }
+
+func (g *GRPC) GetFirstAbortedUploadsInContinuousAborts(
+	ctx context.Context,
+	request *proto.FirstAbortedUploadsInContinuousAbortsRequest,
+) (*proto.FirstAbortedUploadsInContinuousAbortsResponse, error) {
+	g.logger.Infow(
+		"Getting first aborted uploads in a series of continuous aborts",
+		lf.WorkspaceID, request.WorkspaceId,
+	)
```
Comment (achettyiitr) [inline warehouse/api/grpc.go:980]: nitpic: We can use structured logging in here,
```suggestion
	g.logger.Infon(
		"Getting first aborted uploads in a series of continuous aborts",
		obskit.WorkspaceID(request.WorkspaceId),
	)
```

File: warehouse/internal/repo/upload.go [hunk 2]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
...
+	FROM (
+		SELECT 
+			*,
+			ROW_NUMBER() OVER (PARTITION BY destination_id ORDER BY created_at) as row_number
+		FROM wh_uploads_with_last_successful_upload
+		WHERE status = 'aborted'
+		AND (last_successful_upload IS NULL OR created_at > last_successful_upload)
+	) AS q
+	WHERE q.row_number = 1 
+	`, outputColumns, outputColumns)
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1385]:
```suggestion
	stmt := fmt.Sprintf(`
	WITH wh_uploads_with_last_successful_upload AS (
		SELECT 
			%[1]s,
			status,
        	MAX(CASE WHEN status = 'exported_data' THEN created_at ELSE NULL END) OVER (PARTITION BY destination_id) AS last_successful_upload
		FROM 
			`+uploadsTableName+`
		WHERE workspace_id = $1
		AND created_at >= NOW() - INTERVAL '30 day'
	)
	SELECT %[1]s
	FROM (
		SELECT 
			*,
			ROW_NUMBER() OVER (PARTITION BY destination_id ORDER BY created_at) as row_number
		FROM wh_uploads_with_last_successful_upload
		WHERE status = 'aborted'
		AND (last_successful_upload IS NULL OR created_at > last_successful_upload)
	) AS q
	WHERE q.row_number = 1 
	`, outputColumns)
```

File: warehouse/internal/repo/upload.go [hunk 3]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
 	}
 	return nil
 }
+
+func (u *Uploads) GetFirstAbortedUploadsInContinuousAborts(ctx context.Context, workspaceId string) ([]model.FirstAbortedUploadResponse, error) {
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1361]:
```suggestion
func (u *Uploads) GetFirstAbortedUploadsInContinuousAborts(ctx context.Context, workspaceID string) ([]model.FirstAbortedUploadResponse, error) {
```

File: warehouse/internal/repo/upload.go [hunk 4]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
...
+	stmt := fmt.Sprintf(`
+	WITH wh_uploads_with_last_successful_upload AS (
+		SELECT 
+			%s,
+			status,
+        	MAX(CASE WHEN status = 'exported_data' THEN created_at ELSE NULL END) OVER (PARTITION BY destination_id) AS last_successful_upload
+		FROM 
+			`+uploadsTableName+`
+		WHERE workspace_id = $1
+		AND created_at >= NOW() - INTERVAL '30 day'
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1373]: Does interval needs to be hard-coded. I thought control-plane would be sending that information.
```suggestion
		AND created_at >= NOW() - $2 * INTERVAL '1 DAY'
```

File: warehouse/internal/repo/upload.go [hunk 5]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
...
+	)
+	SELECT %s
+	FROM (
+		SELECT 
+			*,
+			ROW_NUMBER() OVER (PARTITION BY destination_id ORDER BY created_at) as row_number
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1379]: nitpick:
```suggestion
			ROW_NUMBER() OVER (PARTITION BY destination_id ORDER BY created_at) AS row_number
```

File: warehouse/internal/repo/upload.go [hunk 6]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
...
+		FROM wh_uploads_with_last_successful_upload
+		WHERE status = 'aborted'
+		AND (last_successful_upload IS NULL OR created_at > last_successful_upload)
+	) AS q
+	WHERE q.row_number = 1
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1384]: nitpick:
```suggestion
	WHERE q.row_number = 1 ;
```

File: warehouse/internal/repo/upload.go [hunk 7]
```diff
@@ -1357,3 +1357,68 @@ func (u *Uploads) update(
...
+		WHERE status = 'aborted'
+		AND (last_successful_upload IS NULL OR created_at > last_successful_upload)
+	) AS q
+	WHERE q.row_number = 1 
+	`, outputColumns, outputColumns)
+
+	rows, err := u.db.QueryContext(ctx, stmt, workspaceId)
+	if err != nil {
+		return nil, fmt.Errorf("first aborted upload in a series of continues aborts info: %w", err)
+	}
```
Comment (achettyiitr) [inline warehouse/internal/repo/upload.go:1390]: We need to close the rows as well in defer.
```suggestion
	rows, err := u.db.QueryContext(ctx, stmt, workspaceID, queryInterval)
	if err != nil {
		return nil, fmt.Errorf("first aborted upload in a series of continues aborts info: %w", err)
	}
	defer func() { _ = rows.Close() }()
```

**Review by achettyiitr — CHANGES_REQUESTED** [review]

File: warehouse/api/grpc.go (hunk 1 above)
Comment (itsmihir) [reply warehouse/api/grpc.go]: obskit.StartTime is not present.

File: warehouse/internal/repo/upload.go (hunk 4 above)
Comment (itsmihir) [reply warehouse/internal/repo/upload.go]: Added a start param

File: warehouse/api/grpc.go (hunk 1 above)
Comment (achettyiitr) [reply warehouse/api/grpc.go]: If something is not there, u can use something liek: logger.NewStringField("xyz", xyz))

File: warehouse/api/grpc.go [hunk 8]
```diff
@@ -969,3 +970,46 @@ func statsInterceptor(statsFactory stats.Stats) grpc.UnaryServerInterceptor {
 		return res, err
 	}
 }
+
+func (g *GRPC) GetFirstAbortedUploadsInContinuousAborts(
```
Comment (cisse21) [inline warehouse/api/grpc.go:974]: [nit-pick]
```suggestion
func (g *GRPC) GetFirstAbortedUploadInContinuousAborts(
```

File: proto/warehouse/warehouse.proto [hunk 9]
```diff
@@ -155,3 +156,21 @@ message RetryFailedBatchesRequest{
 message RetryFailedBatchesResponse{
   int64 retriedSyncsCount = 1;
 }
+
+message FirstAbortedUploadsInContinuousAbortsRequest {
```
Comment (cisse21) [inline proto/warehouse/warehouse.proto:160]:
```suggestion
message FirstAbortedUploadInContinuousAbortsRequest {
```

File: proto/warehouse/warehouse.proto [hunk 10]
```diff
@@ -155,3 +156,21 @@ message RetryFailedBatchesRequest{
...
+message FirstAbortedUploadResponse {
+  int64 id = 1;
+  string source_id = 2;
+  string destination_id = 3;
+  google.protobuf.Timestamp created_at = 4;
+  google.protobuf.Timestamp first_event_at = 5;
+  google.protobuf.Timestamp last_event_at = 6;
+}
+
+message FirstAbortedUploadsInContinuousAbortsResponse {
```
Comment (cisse21) [inline proto/warehouse/warehouse.proto:174]:
```suggestion
message FirstAbortedUploadInContinuousAbortsResponse { 
```

**Review by achettyiitr — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]

**Pull request merged**